sceptre/plan/plan.py

```diff
@@ -6,49 +6,60 @@
 This module implements a SceptrePlan, which is responsible for holding all
 nessessary information for a command to execute.
 """
+import functools
+import itertools
 from os import path, walk
-from typing import Dict
+from typing import Dict, List, Set, Callable, Iterable, Optional
 
-from sceptre.diffing.stack_differ import StackDiff
-from sceptre.exceptions import ConfigFileNotFoundError
 from sceptre.config.graph import StackGraph
 from sceptre.config.reader import ConfigReader
-from sceptre.plan.executor import SceptrePlanExecutor
+from sceptre.context import SceptreContext
+from sceptre.diffing.stack_differ import StackDiff
+from sceptre.exceptions import ConfigFileNotFoundError
 from sceptre.helpers import sceptreise_path
+from sceptre.plan.executor import SceptrePlanExecutor
 from sceptre.stack import Stack
 
 
+def require_resolved(func) -> Callable:
+    @functools.wraps(func)
+    def wrapped(self: "SceptrePlan", *args, **kwargs):
+        if self.launch_order is None:
+            raise RuntimeError(f"You cannot call {func.__name__}() before resolve().")
+        return func(self, *args, **kwargs)
+
+    return wrapped
+
+
 class SceptrePlan(object):
 
-    def __init__(self, context):
+    def __init__(self, context: SceptreContext):
         """
         Intialises a SceptrePlan and generates the Stacks, StackGraph and
         launch order of required.
 
         :param context: A SceptreContext
-        :type sceptre.context.SceptreContext:
         """
         self.context = context
         self.command = None
         self.reverse = None
-        self.launch_order = None
+        self.launch_order: Optional[List[Set[Stack]]] = None
```

Comment:
This blows up when tested with `a1.yaml` stack config file..

```
template:
  path: "noop.yaml"
stack_name: "a1"
launch_action: "skip"
```

```
➜  sceptre launch prod/a1.yaml
* During launch, the following stacks will be skipped, neither created nor deleted:
prod/a1

Do you want to launch 'prod/a1.yaml' [y/N]: y
Traceback (most recent call last):
  File "/Users/zaro0508/.pyenv/versions/py38-sceptredev/bin/sceptre", line 33, in <module>
    sys.exit(load_entry_point('sceptre', 'console_scripts', 'sceptre')())
  File "/Users/zaro0508/.pyenv/versions/3.8.12/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/zaro0508/.pyenv/versions/3.8.12/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/zaro0508/.pyenv/versions/3.8.12/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/zaro0508/.pyenv/versions/3.8.12/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/zaro0508/.pyenv/versions/3.8.12/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/zaro0508/.pyenv/versions/3.8.12/envs/py38-sceptredev/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/zaro0508/work-github/sceptre/sceptre/cli/helpers.py", line 42, in decorated
    return func(*args, **kwargs)
  File "/Users/zaro0508/work-github/sceptre/sceptre/cli/launch.py", line 49, in launch_command
    exit_code = launcher.launch(yes)
  File "/Users/zaro0508/work-github/sceptre/sceptre/cli/launch.py", line 75, in launch
    code = code or self._deploy(deploy_plan)
  File "/Users/zaro0508/work-github/sceptre/sceptre/cli/launch.py", line 172, in _deploy
    result = deploy_plan.launch()
  File "/Users/zaro0508/work-github/sceptre/sceptre/plan/plan.py", line 157, in launch
    return self._execute(*args)
  File "/Users/zaro0508/work-github/sceptre/sceptre/plan/plan.py", line 29, in wrapped
    return func(self, *args, **kwargs)
  File "/Users/zaro0508/work-github/sceptre/sceptre/plan/plan.py", line 56, in _execute
    return executor.execute(*args)
  File "/Users/zaro0508/work-github/sceptre/sceptre/plan/executor.py", line 49, in execute
    with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
  File "/Users/zaro0508/.pyenv/versions/3.8.12/lib/python3.8/concurrent/futures/thread.py", line 141, in __init__
    raise ValueError("max_workers must be greater than 0")
ValueError: max_workers must be greater than 0
```

Also get the same error when a1 stack exists then launch with `launch_action: delete`..

```
template:
  path: "noop.yaml"
stack_name: "a1"
launch_action: "delete"
```

Also get the same error when `launch_action` is set in the stack group config (config.yaml file)

Reply:
Interesting! It looks like it will fail if the launch_action prohibits any stacks from being launched at all. I'll dig in, but I think there's an easy approach here, which would basically just be return early if there's nothing to do.

Reply:
I've have fixed this bug.